gui/RespiraConnectedDevice.h

```diff
@@ -0,0 +1,115 @@
+#include <QSerialPort>
+#include <QThread>
+
+#define INTER_FRAME_TIMEOUT 11
+
+class RespiraConnectedDevice : public ConnectedDevice {
+
+public:
+  RespiraConnectedDevice(QString portName, uint32_t write_timeout_ms,
+                         uint32_t interframe_timeout_ms)
+      : serialPortName(portName), m_writeTimeout_ms(write_timeout_ms),
+        m_interframeTimeout_ms(interframe_timeout_ms) {}
+
+  ~RespiraConnectedDevice() {
+    if (m_serial != nullptr) {
+      m_serial->close();
+    }
+  }
+  void createPort() {
+    if (m_serial != nullptr)
+      return;
+    printf("Socket create\n");
+    m_serial = new QSerialPort();
+    m_serial->setPortName(serialPortName);
+    m_serial->setBaudRate(QSerialPort::Baud115200);
+    m_serial->setDataBits(QSerialPort::Data8);
+    m_serial->setParity(QSerialPort::NoParity);
+    m_serial->setStopBits(QSerialPort::OneStop);
+    m_serial->setFlowControl(QSerialPort::NoFlowControl);
+    if (!m_serial->open(QIODevice::ReadWrite)) {
+      printf("Could not open serial port %s\n",
+             serialPortName.toStdString().c_str());
+    }
+  }
+
+  bool SendGuiStatus(const GuiStatus &gui_status) override {
+    createPort();
+    uint8_t tx_buffer[GuiStatus_size];
+    pb_ostream_t stream = pb_ostream_from_buffer(tx_buffer, sizeof(tx_buffer));
+    if (!pb_encode(&stream, GuiStatus_fields, &gui_status)) {
+      // TODO: Serialization failure; log an error or raise an alert.
+      return false;
+    }
+
+    m_serial->write((const char *)tx_buffer, stream.bytes_written);
+
+    if (!m_serial->waitForBytesWritten(m_writeTimeout_ms)) {
+      // TODO communication failure, port closed? Log an error and raise
+      // an alert
+      return false;
+    }
+    return true;
+  }
+
+#include <chrono>
+  using SteadyClock = std::chrono::steady_clock;
+
+  bool ReceiveControllerStatus(ControllerStatus *controller_status) override {
+    createPort();
+    // Drop everything we have in buffer right now as we are not sure if this is
+    // a full frame or a part of it.
+    // We drop everything that comes in until we get at least half of interframe
+    // silence
+
+    //    if (m_serial->bytesAvailable() > 0) {
+    //      printf("%ld ", SteadyClock::now().time_since_epoch() / 1000000);
+    //      printf("clearing buffer\n");
+
+    //      while (m_serial->waitForReadyRead(m_interframeTimeout_ms / 2)) {
+    //        m_serial->clear(QSerialPort::Input);
+    //      }
+    //    }
+
+    printf("%ld ", SteadyClock::now().time_since_epoch() / 1000000);
+    if (m_serial->waitForReadyRead(m_interframeTimeout_ms)) {
+      QByteArray responseData = m_serial->readAll();
+
+      // continue reading characters until we don't see
+      // any data for long enough time to estimate end of packet silence
+      while (m_serial->waitForReadyRead(m_interframeTimeout_ms / 2)) {
+        responseData += m_serial->readAll();
+      }
+
+        pb_istream_t stream = pb_istream_from_buffer(
```

Comment:
This doesn't look clang-formatted, please make sure you have the pre-commit hooks installed.